File: flutter_app/lib/models/ride.dart
import 'waypoint.dart';

class Ride {
  final String id;
  final String titre;
  final String? description;
  final String typeVehicule;
  final DateTime date;
  final String heure;
  final dynamic lieuDepart;
  final dynamic lieuArrivee;
  final List<Waypoint>? waypoints; // Nouveau système de waypoints
  final double rayon;
  final UserOrganisateur organisateur;
  final String visibilite;
  final String? secretLink; // Lien secret pour les balades secrètes
  final List<UserParticipant> participants;
  final List<RideInvitation>? invitations;
  final List<String> likes;
  final double noteMoyenne;
  final String? url;
  final String? icsUrl;
  final int? totalLikes;
  final bool? hasUserLiked;
  final bool? isOrganizerPremium; // Indique si l'organisateur est premium
  final String status; // Statut de la balade: scheduled, in_progress, completed, cancelled, postponed
  final String? ridingStyle; // Style de conduite: calme, modere, sportif, mixte

  // ========== OUTILS ORGANISATEUR ==========
  final bool requiresApproval; // Validation manuelle des participants
  final List<PendingRequest>? pendingRequests; // Demandes en attente
  final int? maxParticipants; // Limite de participants
  final bool enableWaitlist; // Liste d'attente activée
  final List<WaitlistEntry>? waitlist; // Liste d'attente
  final AutoReminder? autoReminder; // Message automatique avant la balade
  final RideRecurrence? recurrence; // Balades récurrentes

  Ride({
    required this.id,
    required this.titre,
    this.description,
    required this.typeVehicule,
    required this.date,
    required this.heure,
    required this.lieuDepart,
    required this.lieuArrivee,
    this.waypoints,
    required this.rayon,
    required this.organisateur,
    required this.visibilite,
    this.secretLink,
    required this.participants,
    this.invitations,
    required this.likes,
    required this.noteMoyenne,
    this.url,
    this.icsUrl,
    this.totalLikes,
    this.hasUserLiked,
    this.isOrganizerPremium,
    this.status = 'scheduled',
    this.ridingStyle,
    this.requiresApproval = false,
    this.pendingRequests,
    this.maxParticipants,
    this.enableWaitlist = false,
    this.waitlist,
    this.autoReminder,
    this.recurrence,
  });

  factory Ride.fromJson(Map<String, dynamic> json) {
    return Ride(
      id: json['id'] ?? json['_id'] ?? '',
      titre: json['titre'] ?? '',
      description: json['description'],
      typeVehicule: json['typeVehicule'] ?? '',
      date: json['date'] != null 
          ? DateTime.parse(json['date'])
          : DateTime.now(),
      heure: json['heure'] ?? '',
      lieuDepart: json['lieuDepart'],
      lieuArrivee: json['lieuArrivee'],
      waypoints: json['waypoints'] != null
          ? (json['waypoints'] as List)
              .map((w) => Waypoint.fromJson(w))
              .toList()
          : null,
      rayon: (json['rayon'] ?? 0).toDouble(),
      organisateur: UserOrganisateur.fromJson(
        json['organisateur'] is Map 
            ? json['organisateur'] 
            : {'_id': json['organisateur']}
      ),
      visibilite: json['visibilite'] ?? 'publique',
      secretLink: json['secretLink'],
      participants: json['participants'] != null
          ? (json['participants'] as List)
              .map((p) {
                // Gérer la nouvelle structure avec userId comme sous-document
                if (p is Map && p.containsKey('userId')) {
                  // Si userId est un objet (populé), utiliser ses données
                  final userId = p['userId'];
                  if (userId is Map) {
                    return UserParticipant.fromJson({
                      'id': userId['_id'] ?? userId['id'] ?? '',
                      'firstName': userId['firstName'],
                      'lastName': userId['lastName'],
                      'pseudo': userId['pseudo'],
                    });
                  } else {
                    // Si userId est juste un ID, utiliser l'ID
                    return UserParticipant.fromJson({
                      'id': userId.toString(),
                    });
                  }
                }
                // Ancienne structure (participant direct)
                return UserParticipant.fromJson(p);
              })
              .toList()
          : [],
      invitations: json['invitations'] != null
          ? (json['invitations'] as List)
              .map((inv) => RideInvitation.fromJson(inv))
              .toList()
          : null,
      likes: json['likes'] != null 
          ? List<String>.from(json['likes'].map((l) => l.toString()))
          : [],
      noteMoyenne: (json['noteMoyenne'] ?? 0).toDouble(),
      url: json['url'],
      icsUrl: json['icsUrl'],
      totalLikes: json['totalLikes'],
      hasUserLiked: json['hasUserLiked'],
      isOrganizerPremium: json['isOrganizerPremium'] ?? false,
      status: json['status'] ?? 'scheduled',
      ridingStyle: json['ridingStyle'],
      // Outils organisateur
      requiresApproval: json['requiresApproval'] ?? false,
      pendingRequests: json['pendingRequests'] != null
          ? (json['pendingRequests'] as List)
              .map((r) => PendingRequest.fromJson(r))
              .toList()
          : null,
      maxParticipants: json['maxParticipants'],
      enableWaitlist: json['enableWaitlist'] ?? false,
      waitlist: json['waitlist'] != null
          ? (json['waitlist'] as List)
              .map((w) => WaitlistEntry.fromJson(w))
              .toList()
          : null,
      autoReminder: json['autoReminder'] != null
          ? AutoReminder.fromJson(json['autoReminder'])
          : null,
      recurrence: json['recurrence'] != null
          ? RideRecurrence.fromJson(json['recurrence'])
          : null,
    );
  }

  bool get isLikedByCurrentUser => false; // Sera géré par le state
  bool get isParticipant => false; // Sera géré par le state
  
  // Helpers pour les outils organisateur
  bool get hasOrganizerTools => requiresApproval || maxParticipants != null || 
      enableWaitlist || (autoReminder?.enabled ?? false) || (recurrence?.enabled ?? false);
  
  int get pendingRequestsCount => pendingRequests?.length ?? 0;
  int get waitlistCount => waitlist?.length ?? 0;
  bool get isFull => maxParticipants != null && participants.length >= maxParticipants!;
}

class UserOrganisateur {
  final String id;
  final String? firstName;
  final String? lastName;
  final String? pseudo;
  final String email;

  UserOrganisateur({
    required this.id,
    this.firstName,
    this.lastName,
    this.pseudo,
    required this.email,
  });

  factory UserOrganisateur.fromJson(Map<String, dynamic> json) {
    return UserOrganisateur(
      id: json['id'] ?? json['_id'] ?? '',
      firstName: json['firstName'],
      lastName: json['lastName'],
      pseudo: json['pseudo'],
      email: json['email'] ?? '',
    );
  }

  String get displayName {
    if (firstName != null && lastName != null) {
      return '$firstName $lastName';
    } else if (pseudo != null) {
      return pseudo!;
    }
    return email;
  }
}

class UserParticipant {
  final String id;
  final String? firstName;
  final String? lastName;
  final String? pseudo;

  UserParticipant({
    required this.id,
    this.firstName,
    this.lastName,
    this.pseudo,
  });

  factory UserParticipant.fromJson(Map<String, dynamic> json) {
    return UserParticipant(
      id: json['id'] ?? json['_id'] ?? '',
      firstName: json['firstName'],
      lastName: json['lastName'],
      pseudo: json['pseudo'],
    );
  }

  String get displayName {
    if (firstName != null && lastName != null) {
      return '$firstName $lastName';
    } else if (pseudo != null) {
      return pseudo!;
    }
    return 'Utilisateur';
  }
}

class RideInvitation {
  final String userId;
  final String? firstName;
  final String? lastName;
  final String? pseudo;
  final String status; // 'pending', 'accepted', 'declined'
  final DateTime? invitedAt;
  final DateTime? respondedAt;

  RideInvitation({
    required this.userId,
    this.firstName,
    this.lastName,
    this.pseudo,
    required this.status,
    this.invitedAt,
    this.respondedAt,
  });

  factory RideInvitation.fromJson(Map<String, dynamic> json) {
    // Gérer le cas où userId est un objet (populé)
    String userIdStr;
    String? firstName;
    String? lastName;
    String? pseudo;

    if (json['userId'] is Map) {
      final user = json['userId'];
      userIdStr = user['_id'] ?? user['id'] ?? '';
      firstName = user['firstName'];
      lastName = user['lastName'];
      pseudo = user['pseudo'];
    } else {
      userIdStr = json['userId']?.toString() ?? '';
    }

    return RideInvitation(
      userId: userIdStr,
      firstName: firstName ?? json['firstName'],
      lastName: lastName ?? json['lastName'],
      pseudo: pseudo ?? json['pseudo'],
      status: json['status'] ?? 'pending',
      invitedAt: json['invitedAt'] != null ? DateTime.parse(json['invitedAt']) : null,
      respondedAt: json['respondedAt'] != null ? DateTime.parse(json['respondedAt']) : null,
    );
  }

  String get displayName {
    if (firstName != null && lastName != null) {
      return '$firstName $lastName';
    } else if (pseudo != null) {
      return pseudo!;
    }
    return 'Utilisateur';
  }
}

// ========== CLASSES POUR LES OUTILS ORGANISATEUR ==========

/// Demande de participation en attente d'approbation
class PendingRequest {
  final String userId;
  final String? firstName;
  final String? lastName;
  final String? pseudo;
  final String? avatarUrl;
  final String? vehicleId;
  final String? vehicleNickname;
  final DateTime requestedAt;
  final String? message;

  PendingRequest({
    required this.userId,
    this.firstName,
    this.lastName,
    this.pseudo,
    this.avatarUrl,
    this.vehicleId,
    this.vehicleNickname,
    required this.requestedAt,
    this.message,
  });

  factory PendingRequest.fromJson(Map<String, dynamic> json) {
    String parsedUserId;
    String? parsedFirstName;
    String? parsedLastName;
    String? parsedPseudo;
    String? parsedAvatarUrl;

    if (json['userId'] is Map) {
      final user = json['userId'];
      parsedUserId = user['_id'] ?? user['id'] ?? '';
      parsedFirstName = user['firstName'];
      parsedLastName = user['lastName'];
      parsedPseudo = user['pseudo'];
      parsedAvatarUrl = user['avatarUrl'];
    } else {
      parsedUserId = json['userId']?.toString() ?? '';
    }

    String? parsedVehicleId;
    String? parsedVehicleNickname;
    if (json['vehicleId'] is Map) {
      final vehicle = json['vehicleId'];
      parsedVehicleId = vehicle['_id'] ?? vehicle['id'];
      parsedVehicleNickname = vehicle['nickname'] ?? '${vehicle['make']} ${vehicle['model']}';
    } else {
      parsedVehicleId = json['vehicleId']?.toString();
    }

    return PendingRequest(
      userId: parsedUserId,
      firstName: parsedFirstName,
      lastName: parsedLastName,
      pseudo: parsedPseudo,
      avatarUrl: parsedAvatarUrl,
      vehicleId: parsedVehicleId,
      vehicleNickname: parsedVehicleNickname,
      requestedAt: json['requestedAt'] != null 
          ? DateTime.parse(json['requestedAt']) 
          : DateTime.now(),
      message: json['message'],
    );
  }

  String get displayName {
    if (firstName != null && lastName != null) {
      return '$firstName $lastName';
    } else if (pseudo != null) {
      return pseudo!;
    }
    return 'Utilisateur';
  }
}

/// Entrée dans la liste d'attente
class WaitlistEntry {
  final String userId;
  final String? firstName;
  final String? lastName;
  final String? pseudo;
  final String? avatarUrl;
  final String? vehicleId;
  final String? vehicleNickname;
  final DateTime addedAt;
  final int position;

  WaitlistEntry({
    required this.userId,
    this.firstName,
    this.lastName,
    this.pseudo,
    this.avatarUrl,
    this.vehicleId,
    this.vehicleNickname,
    required this.addedAt,
    required this.position,
  });

  factory WaitlistEntry.fromJson(Map<String, dynamic> json) {
    String userId;
    String? firstName;
    String? lastName;
    String? pseudo;
    String? avatarUrl;

    if (json['userId'] is Map) {
      final user = json['userId'];
      userId = user['_id'] ?? user['id'] ?? '';
      firstName = user['firstName'];
      lastName = user['lastName'];
      pseudo = user['pseudo'];
      avatarUrl = user['avatarUrl'];
    } else {
      userId = json['userId']?.toString() ?? '';
    }

    String? vehicleId;
    String? vehicleNickname;
    if (json['vehicleId'] is Map) {
      final vehicle = json['vehicleId'];
      vehicleId = vehicle['_id'] ?? vehicle['id'];
      vehicleNickname = vehicle['nickname'] ?? '${vehicle['make']} ${vehicle['model']}';
    } else {
      vehicleId = json['vehicleId']?.toString();
    }

    return WaitlistEntry(
      userId: userId,
      firstName: firstName,
      lastName: lastName,
      pseudo: pseudo,
      avatarUrl: avatarUrl,
      vehicleId: vehicleId,
      vehicleNickname: vehicleNickname,
      addedAt: json['addedAt'] != null 
          ? DateTime.parse(json['addedAt']) 
          : DateTime.now(),
      position: json['position'] ?? 0,
    );
  }

  String get displayName {
    if (firstName != null && lastName != null) {
      return '$firstName $lastName';
    } else if (pseudo != null) {
      return pseudo!;
    }
    return 'Utilisateur';
  }
}

/// Configuration du rappel automatique
class AutoReminder {
  final bool enabled;
  final int hoursBefore;
  final String? message;
  final DateTime? sentAt;

  AutoReminder({
    required this.enabled,
    required this.hoursBefore,
    this.message,
    this.sentAt,
  });

  factory AutoReminder.fromJson(Map<String, dynamic> json) {
    return AutoReminder(
      enabled: json['enabled'] ?? false,
      hoursBefore: json['hoursBefore'] ?? 24,
      message: json['message'],
      sentAt: json['sentAt'] != null ? DateTime.parse(json['sentAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'enabled': enabled,
      'hoursBefore': hoursBefore,
      'message': message,
    };
  }
}

/// Configuration de la récurrence
class RideRecurrence {
  final bool enabled;
  final String frequency; // 'weekly', 'biweekly', 'monthly'
  final int? dayOfWeek; // 0 = dimanche, 1 = lundi, ..., 6 = samedi
  final DateTime? endDate;
  final String? parentRideId;
  final DateTime? nextOccurrence;

  RideRecurrence({
    required this.enabled,
    required this.frequency,
    this.dayOfWeek,
    this.endDate,
    this.parentRideId,
    this.nextOccurrence,
  });

  factory RideRecurrence.fromJson(Map<String, dynamic> json) {
    return RideRecurrence(
      enabled: json['enabled'] ?? false,
      frequency: json['frequency'] ?? 'weekly',
      dayOfWeek: json['dayOfWeek'],
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
      parentRideId: json['parentRideId']?.toString(),
      nextOccurrence: json['nextOccurrence'] != null 
          ? DateTime.parse(json['nextOccurrence']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'enabled': enabled,
      'frequency': frequency,
      'dayOfWeek': dayOfWeek,
      'endDate': endDate?.toIso8601String(),
    };
  }

  String get frequencyLabel {
    switch (frequency) {
      case 'weekly':
        return 'Hebdomadaire';
      case 'biweekly':
        return 'Toutes les 2 semaines';
      case 'monthly':
        return 'Mensuelle';
      default:
        return frequency;
    }
  }

  String get dayOfWeekLabel {
    if (dayOfWeek == null) return '';
    const days = ['Dimanche', 'Lundi', 'Mardi', 'Mercredi', 'Jeudi', 'Vendredi', 'Samedi'];
    return days[dayOfWeek!];
  }
}

File: src/models/Ride.js
const mongoose = require('mongoose');

const rideSchema = new mongoose.Schema({
  titre: {
    type: String,
    required: [true, 'Le titre est requis'],
    trim: true,
    maxlength: [200, 'Le titre ne peut pas dépasser 200 caractères']
  },
  description: {
    type: String,
    trim: true,
    maxlength: [2000, 'La description ne peut pas dépasser 2000 caractères']
  },
  typeVehicule: {
    type: String,
    enum: ['moto', 'voiture'],
    required: [true, 'Le type de véhicule est requis']
  },
  date: {
    type: Date,
    required: [true, 'La date est requise']
  },
  heure: {
    type: String,
    required: [true, 'L\'heure est requise'],
    match: [/^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$/, 'Format d\'heure invalide (HH:MM)']
  },
  lieuDepart: {
    type: mongoose.Schema.Types.Mixed,
    required: [true, 'Le lieu de départ est requis']
    // Peut être une string ou un objet GeoJSON
  },
  lieuArrivee: {
    type: mongoose.Schema.Types.Mixed,
    required: [true, 'Le lieu d\'arrivée est requis']
    // Peut être une string ou un objet GeoJSON
  },
  // Nouveau système de waypoints (comme Liberty Rider)
  waypoints: [{
    _id: {
      type: mongoose.Schema.Types.ObjectId,
      auto: true
    },
    type: {
      type: String,
      enum: ['depart', 'checkpoint', 'arrivee'],
      required: true
    },
    address: {
      type: String,
      required: true
    },
    coordinates: {
      type: {
        type: String,
        enum: ['Point'],
        default: 'Point'
      },
      coordinates: {
        type: [Number], // [longitude, latitude]
        required: true,
        validate: {
          validator: function(coords) {
            return coords.length === 2 && 
                   coords[0] >= -180 && coords[0] <= 180 && 
                   coords[1] >= -90 && coords[1] <= 90;
          },
          message: 'Les coordonnées doivent être [longitude, latitude]'
        }
      }
    },
    order: {
      type: Number,
      required: true,
      min: 0
    },
    // NOUVEAUX CHAMPS - Types de waypoints avancés
    waypointType: {
      type: String,
      enum: ['normal', 'fuel', 'coffee', 'danger', 'viewpoint'],
      default: 'normal'
    },
    isMandatoryStop: {
      type: Boolean,
      default: false
    },
    note: {
      type: String,
      maxlength: [500, 'La note ne peut pas dépasser 500 caractères'],
      default: null
    },
    createdBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  localisation: {
    type: {
      type: String,
      enum: ['Point'],
      default: 'Point'
    },
    coordinates: {
      type: [Number], // [longitude, latitude]
      required: false,
      validate: {
        validator: function(coords) {
          return !coords || (coords.length === 2 && 
                 coords[0] >= -180 && coords[0] <= 180 && 
                 coords[1] >= -90 && coords[1] <= 90);
        },
        message: 'Les coordonnées doivent être [longitude, latitude] avec longitude entre -180 et 180, latitude entre -90 et 90'
      }
    }
  },
  rayon: {
    type: Number,
    required: false, // Le rayon n'est plus requis, il est utilisé uniquement pour la recherche
    min: [0, 'Le rayon doit être positif'],
    default: 0
  },
  organisateur: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'L\'organisateur est requis']
  },
  // Association avec un groupe (optionnel, rétrocompatible)
  groupId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Group',
    default: null,
    index: true
  },
  visibilite: {
    type: String,
    enum: ['privee', 'publique', 'secrete'],
    default: 'publique'
  },
  
  // Lien secret pour les balades secrètes (généré automatiquement)
  secretLink: {
    type: String,
  },
  participants: [{
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    vehicleId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Vehicle',
      default: null // Véhicule avec lequel le participant effectue la balade (optionnel)
    },
    // Informations de ponctualité
    arrivalTime: {
      type: Date,
      default: null // Heure à laquelle le participant a indiqué son arrivée
    },
    isOnTime: {
      type: Boolean,
      default: null // null = non validé (considéré comme à l'heure par défaut), true = à l'heure, false = en retard
    },
    validatedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null // Qui a validé/invalidé (généralement l'organisateur)
    },
    validatedAt: {
      type: Date,
      default: null // Quand la validation a eu lieu
    }
  }],
  invitations: [{
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    status: {
      type: String,
      enum: ['pending', 'accepted', 'declined'],
      default: 'pending'
    },
    invitedAt: {
      type: Date,
      default: Date.now
    },
    respondedAt: {
      type: Date,
      default: null
    }
  }],
  likes: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  notes: [{
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    note: {
      type: Number,
      required: true,
      min: [0, 'La note doit être entre 0 et 5'],
      max: [5, 'La note doit être entre 0 et 5']
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  noteMoyenne: {
    type: Number,
    default: 0,
    min: 0,
    max: 5
  },
  // Statut de la balade
  status: {
    type: String,
    enum: ['scheduled', 'in_progress', 'completed', 'cancelled', 'postponed'],
    default: 'scheduled',
    index: true
  },
  // Style de conduite
  ridingStyle: {
    type: String,
    enum: ['calme', 'modere', 'sportif', 'mixte'],
    default: null
  },
  // ========== ANNULATION / REPORT ==========
  // Annulation
  cancellation: {
    cancelledAt: {
      type: Date,
      default: null
    },
    cancelledBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null
    },
    cancelReasonCode: {
      type: String,
      enum: ['WEATHER', 'MECHANICAL', 'ROAD_CLOSED', 'LOW_PARTICIPATION', 'OTHER'],
      default: null
    },
    cancelReasonText: {
      type: String,
      maxlength: [500, 'Le texte de raison ne peut pas dépasser 500 caractères'],
      default: null
    }
  },
  // Report (reporté sans nouvelle date)
  postponement: {
    postponedAt: {
      type: Date,
      default: null
    },
    postponedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null
    },
    postponeReasonCode: {
      type: String,
      enum: ['WEATHER', 'MECHANICAL', 'ROAD_CLOSED', 'LOW_PARTICIPATION', 'OTHER'],
      default: null
    },
    postponeReasonText: {
      type: String,
      maxlength: [500, 'Le texte de raison ne peut pas dépasser 500 caractères'],
      default: null
    },
    newDateTime: {
      type: Date,
      default: null // Date/heure proposée pour le report (optionnel)
    }
  },
  // Reprogrammation (duplication)
  originalRideId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Ride',
    default: null // Si cette balade est une reprogrammation, référence à la balade originale
  },
  reprogrammedToRideId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Ride',
    default: null // Si cette balade a été reprogrammée, référence à la nouvelle balade
  },

  // ========== OUTILS ORGANISATEUR ==========
  
  // 1) Validation manuelle des participants
  requiresApproval: {
    type: Boolean,
    default: false // Si true, les demandes doivent être approuvées par l'organisateur
  },
  
  // Demandes de participation en attente d'approbation
  pendingRequests: [{
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    vehicleId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Vehicle',
      default: null
    },
    requestedAt: {
      type: Date,
      default: Date.now
    },
    message: {
      type: String,
      maxlength: [500, 'Le message ne peut pas dépasser 500 caractères'],
      default: null
    }
  }],
  
  // 2) Limite de participants + liste d'attente
  maxParticipants: {
    type: Number,
    default: null, // null = illimité
    min: [1, 'La limite doit être au moins 1']
  },
  
  enableWaitlist: {
    type: Boolean,
    default: false // Si true, les participants au-delà de la limite sont mis en liste d'attente
  },
  
  waitlist: [{
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    vehicleId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Vehicle',
      default: null
    },
    addedAt: {
      type: Date,
      default: Date.now
    },
    position: {
      type: Number,
      required: true
    }
  }],
  
  // 3) Message automatique avant la balade
  autoReminder: {
    enabled: {
      type: Boolean,
      default: false
    },
    hoursBefore: {
      type: Number,
      default: 24, // Par défaut 24h avant
      min: [1, 'Le rappel doit être au moins 1 heure avant'],
      max: [168, 'Le rappel ne peut pas être plus de 7 jours avant']
    },
    message: {
      type: String,
      maxlength: [1000, 'Le message ne peut pas dépasser 1000 caractères'],
      default: null // Si null, un message par défaut sera utilisé
    },
    sentAt: {
      type: Date,
      default: null // Date à laquelle le rappel a été envoyé
    }
  },
  
  // 4) Balades récurrentes
  recurrence: {
    enabled: {
      type: Boolean,
      default: false
    },
    frequency: {
      type: String,
      enum: ['weekly', 'biweekly', 'monthly'],
      default: 'weekly'
    },
    dayOfWeek: {
      type: Number, // 0 = dimanche, 1 = lundi, ..., 6 = samedi
      min: 0,
      max: 6,
      default: null
    },
    endDate: {
      type: Date,
      default: null // Date de fin de la récurrence (null = infini)
    },
    parentRideId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Ride',
      default: null // Pour les balades créées automatiquement, référence à la balade parente
    },
    nextOccurrence: {
      type: Date,
      default: null // Prochaine occurrence à créer
    }
  },
  // Événements de la balade (pour mode live)
  rideEvents: [{
    type: {
      type: String,
      enum: ['started', 'paused', 'resumed', 'incident', 'completed', 'cancelled', 'participant_joined', 'participant_left'],
      required: true
    },
    timestamp: {
      type: Date,
      default: Date.now
    },
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    details: {
      location: {
        type: {
          type: String,
          enum: ['Point'],
          default: 'Point'
        },
        coordinates: {
          type: [Number], // [longitude, latitude]
          default: null
        }
      },
      description: {
        type: String,
        trim: true,
        maxlength: [500, 'La description ne peut pas dépasser 500 caractères']
      }
    }
  }]
}, {
  timestamps: true
});

// Index pour améliorer les performances de recherche
rideSchema.index({ organisateur: 1 });
rideSchema.index({ date: 1 });
rideSchema.index({ typeVehicule: 1 });
rideSchema.index({ visibilite: 1 });
rideSchema.index({ participants: 1 });
rideSchema.index({ 'participants.vehicleId': 1 }); // Pour rechercher les balades par véhicule
rideSchema.index({ status: 1, date: 1 }); // Pour les requêtes par statut
rideSchema.index({ ridingStyle: 1 }); // Pour les filtres par style
// Index pour calendar groupe
rideSchema.index({ groupId: 1, date: 1 }); // Calendar queries
rideSchema.index({ groupId: 1, status: 1, date: 1 }); // Filtered calendar
// Index compound pour requêtes fréquentes (listing avec filtres)
rideSchema.index({ typeVehicule: 1, date: 1, visibilite: 1 });
rideSchema.index({ organisateur: 1, date: -1 }); // Mes balades triées par date
rideSchema.index({ 'participants.userId': 1, date: -1 }); // Balades où je participe
rideSchema.index({ createdAt: -1 }); // Tri par date de création
rideSchema.index({ 'invitations.userId': 1 }); // Pour rechercher les invitations par utilisateur
rideSchema.index({ 'invitations.status': 1 }); // Pour filtrer par statut d'invitation
// Index géospatial 2dsphere pour les requêtes de proximité
rideSchema.index({ localisation: '2dsphere' });
// Index pour les outils organisateur
rideSchema.index({ 'pendingRequests.userId': 1 });
rideSchema.index({ 'waitlist.userId': 1 });
rideSchema.index({ 'recurrence.parentRideId': 1 });
rideSchema.index({ 'recurrence.nextOccurrence': 1 });
rideSchema.index({ 'autoReminder.enabled': 1, 'autoReminder.sentAt': 1, date: 1 });
rideSchema.index({ secretLink: 1 }, { unique: true, sparse: true });

// Méthode pour calculer la note moyenne
rideSchema.methods.calculateAverageRating = function() {
  if (this.notes.length === 0) {
    this.noteMoyenne = 0;
    return;
  }
  
  const sum = this.notes.reduce((acc, note) => acc + note.note, 0);
  this.noteMoyenne = Math.round((sum / this.notes.length) * 10) / 10; // Arrondi à 1 décimale
};

// Middleware pre-save pour calculer la note moyenne
rideSchema.pre('save', function(next) {
  if (this.isModified('notes')) {
    this.calculateAverageRating();
  }
  next();
});

module.exports = mongoose.model('Ride', rideSchema);

